### PR TITLE
♻️ [RUMF-1083] rework session cookie cache

### PR DESCRIPTION
**Pull request by bcaudan**
## Motivation

Second part of session management rework, cf #1166 

Next:

- Introduce session context history

## Changes

- Remove cookie cache utility
- Expose session through in memory cache
- Watch cookie regularly to expire session 

## Testing

- [x] Local
- [x] Staging
- [x] Unit
- [ ] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/core/src/domain/session/sessionManagement.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ export function startSessionManagement<TrackingType extends string>(
 ): Session<TrackingType> {
   tryOldCookiesMigration(options)
   const sessionStore = startSessionStore(options, productKey, computeSessionState)
+  stopCallbacks.push(() => sessionStore.stop())
```
Comment (amortemousque) [inline packages/core/src/domain/session/sessionManagement.ts:23]: Can we move `stopCallbacks` declaration higher up in the file?

**Review by amortemousque — APPROVED** [review]

File: packages/core/src/domain/session/sessionStore.ts [hunk 2]
```diff
@@ -31,47 +32,65 @@ export function startSessionStore<TrackingType extends string>(
...
+
+  function sessionCacheOutdated(cookieSession: SessionState) {
+    return sessionCache.id !== cookieSession.id
+  }
+
+  function clearSessionCache() {
+    sessionCache = {}
   }
 
   function retrieveSession() {
```
Comment (BenoitZugmeyer) [inline packages/core/src/domain/session/sessionStore.ts:82]: I would name it `retrieveSessionCache` to be unified with other namings and to disambiguate with the other `retrieveSession` function below

File: packages/core/src/domain/session/sessionStore.ts [hunk 3]
```diff
@@ -31,47 +32,65 @@ export function startSessionStore<TrackingType extends string>(
...
 
-      // If the session id has changed, notify that the session has been renewed
-      if (isTracked && inMemorySession.id !== cookieSession.id) {
-        inMemorySession = { ...cookieSession }
+      if (isTracked && sessionCacheOutdated(cookieSession)) {
+        sessionCache = { ...cookieSession }
         renewObservable.notify()
       }
-      inMemorySession = { ...cookieSession }
+      sessionCache = { ...cookieSession }
```
[the remark this answers is not in the document]
Comment (bcaudan) [reply packages/core/src/domain/session/sessionStore.ts:53]: I have started to tweak this part in the next PR, I am in favor of postponing this discussion for the next one then.

[the remark this answers is not in the document]
Comment (BenoitZugmeyer) [reply packages/core/src/domain/session/sessionStore.ts:53]: Okay!

File: packages/core/src/domain/session/sessionStore.ts [hunk 4]
```diff
@@ -31,47 +32,65 @@ export function startSessionStore<TrackingType extends string>(
   computeSessionState: (rawTrackingType?: string) => { trackingType: TrackingType; isTracked: boolean }
 ): SessionStore {
   const renewObservable = new Observable<void>()
-  const sessionCookie = cacheCookieAccess(SESSION_COOKIE_NAME, options)
-  let inMemorySession = retrieveActiveSession(sessionCookie)
+  let sessionCache: SessionState = retrieveActiveSession(options)
```
[the remark this answers is not in the document]
Comment (bcaudan) [reply packages/core/src/domain/session/sessionStore.ts:35]: any issue with doing two `getCookie` at start?
not having the cache warm at start would complexify a bit all the things

File: packages/core/src/domain/session/sessionStore.ts (hunk 2 above)
Comment (bcaudan) [reply packages/core/src/domain/session/sessionStore.ts:82]: Not in favor of exposing the fact that we use a cache to consumers.
I am going with `retrieveSession` -> `getSession`

Comment (BenoitZugmeyer) [reply packages/core/src/domain/session/sessionStore.ts]: Sure, me neither. What about renaming the function internally to harmonize all cache-related functions namings, but still return it as  `retrieveSession`?

Comment (bcaudan) [reply packages/core/src/domain/session/sessionStore.ts]: I have removed the internal function in f353f77.
let me know what you think

File: packages/core/src/domain/session/sessionStore.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (BenoitZugmeyer) [reply packages/core/src/domain/session/sessionStore.ts:35]: Appart from a small performance impact, it should be fine. If it complexify the code too much, let's keep it this way.

[the remark this answers is not in the document]
Comment (bcaudan) [reply packages/core/src/domain/session/sessionStore.ts:35]: IMO, as long as our cookie accesses are constant, we are fine.
Handling sessionCache undefined case in every session store function would be a shame.

**Review by BenoitZugmeyer — APPROVED** [review]

**Pull request merged**